#!/usr/bin/env python3
"""
Teste abrangente das opera√ß√µes CRUD implementadas no gRPC
"""

import grpc
import streaming_pb2
import streaming_pb2_grpc
import sys
import time

def test_grpc_crud_operations():
    """Testa todas as opera√ß√µes CRUD implementadas"""
    
    print("üß™ Iniciando testes das opera√ß√µes CRUD gRPC...")
    
    try:
        # Conectar ao servidor gRPC
        channel = grpc.insecure_channel('localhost:50051')
        stub = streaming_pb2_grpc.StreamingServiceStub(channel)
        
        # Testar se o servidor est√° rodando
        try:
            request = streaming_pb2.google_dot_protobuf_dot_empty__pb2.Empty()
            stub.ListarUsuarios(request)
            print("‚úÖ Conex√£o com servidor gRPC estabelecida")
        except grpc.RpcError as e:
            print(f"‚ùå Erro ao conectar com servidor gRPC: {e}")
            print("üí° Certifique-se de que o servidor gRPC est√° rodando na porta 50051")
            return False
        
        # === TESTES DE USU√ÅRIOS ===
        print("\nüìÅ Testando opera√ß√µes de Usu√°rios:")
        
        # 1. Criar usu√°rio
        print("  üî∏ Testando cria√ß√£o de usu√°rio...")
        try:
            criar_request = streaming_pb2.CriarUsuarioRequest(
                nome="Jo√£o Teste",
                idade=30
            )
            novo_usuario = stub.CriarUsuario(criar_request)
            print(f"    ‚úÖ Usu√°rio criado: {novo_usuario.nome} (ID: {novo_usuario.id_usuario})")
        except grpc.RpcError as e:
            print(f"    ‚ùå Erro ao criar usu√°rio: {e.details()}")
            return False
        
        # 2. Obter usu√°rio
        print("  üî∏ Testando obten√ß√£o de usu√°rio...")
        try:
            obter_request = streaming_pb2.IdRequest(id="user_1")
            usuario = stub.ObterUsuario(obter_request)
            print(f"    ‚úÖ Usu√°rio obtido: {usuario.nome}")
        except grpc.RpcError as e:
            print(f"    ‚ùå Erro ao obter usu√°rio: {e.details()}")
        
        # 3. Atualizar usu√°rio
        print("  üî∏ Testando atualiza√ß√£o de usu√°rio...")
        try:
            atualizar_request = streaming_pb2.AtualizarUsuarioRequest(
                id_usuario="user_1",
                nome="Jo√£o Silva Atualizado",
                idade=31
            )
            usuario_atualizado = stub.AtualizarUsuario(atualizar_request)
            print(f"    ‚úÖ Usu√°rio atualizado: {usuario_atualizado.nome}")
        except grpc.RpcError as e:
            print(f"    ‚ùå Erro ao atualizar usu√°rio: {e.details()}")
        
        # 4. Deletar usu√°rio
        print("  üî∏ Testando dele√ß√£o de usu√°rio...")
        try:
            deletar_request = streaming_pb2.IdRequest(id="user_1")
            resultado = stub.DeletarUsuario(deletar_request)
            print(f"    ‚úÖ Usu√°rio deletado: {resultado.success} - {resultado.message}")
        except grpc.RpcError as e:
            print(f"    ‚ùå Erro ao deletar usu√°rio: {e.details()}")
        
        # === TESTES DE M√öSICAS ===
        print("\nüéµ Testando opera√ß√µes de M√∫sicas:")
        
        # 1. Criar m√∫sica
        print("  üî∏ Testando cria√ß√£o de m√∫sica...")
        try:
            criar_musica_request = streaming_pb2.CriarMusicaRequest(
                nome="M√∫sica Teste",
                artista="Artista Teste",
                duracao_segundos=180
            )
            nova_musica = stub.CriarMusica(criar_musica_request)
            print(f"    ‚úÖ M√∫sica criada: {nova_musica.nome} - {nova_musica.artista}")
        except grpc.RpcError as e:
            print(f"    ‚ùå Erro ao criar m√∫sica: {e.details()}")
        
        # 2. Obter m√∫sica
        print("  üî∏ Testando obten√ß√£o de m√∫sica...")
        try:
            obter_musica_request = streaming_pb2.IdRequest(id="musica_1")
            musica = stub.ObterMusica(obter_musica_request)
            print(f"    ‚úÖ M√∫sica obtida: {musica.nome}")
        except grpc.RpcError as e:
            print(f"    ‚ùå Erro ao obter m√∫sica: {e.details()}")
        
        # 3. Atualizar m√∫sica
        print("  üî∏ Testando atualiza√ß√£o de m√∫sica...")
        try:
            atualizar_musica_request = streaming_pb2.AtualizarMusicaRequest(
                id_musica="musica_1",
                nome="M√∫sica Atualizada",
                artista="Artista Atualizado",
                duracao_segundos=200
            )
            musica_atualizada = stub.AtualizarMusica(atualizar_musica_request)
            print(f"    ‚úÖ M√∫sica atualizada: {musica_atualizada.nome}")
        except grpc.RpcError as e:
            print(f"    ‚ùå Erro ao atualizar m√∫sica: {e.details()}")
        
        # 4. Deletar m√∫sica
        print("  üî∏ Testando dele√ß√£o de m√∫sica...")
        try:
            deletar_musica_request = streaming_pb2.IdRequest(id="musica_1")
            resultado = stub.DeletarMusica(deletar_musica_request)
            print(f"    ‚úÖ M√∫sica deletada: {resultado.success} - {resultado.message}")
        except grpc.RpcError as e:
            print(f"    ‚ùå Erro ao deletar m√∫sica: {e.details()}")
        
        # === TESTES DE PLAYLISTS ===
        print("\nüìã Testando opera√ß√µes de Playlists:")
        
        # 1. Criar playlist
        print("  üî∏ Testando cria√ß√£o de playlist...")
        try:
            criar_playlist_request = streaming_pb2.CriarPlaylistRequest(
                nome="Playlist Teste",
                id_usuario="user_1",
                musicas=["musica_1", "musica_2"]
            )
            nova_playlist = stub.CriarPlaylist(criar_playlist_request)
            print(f"    ‚úÖ Playlist criada: {nova_playlist.nome}")
        except grpc.RpcError as e:
            print(f"    ‚ùå Erro ao criar playlist: {e.details()}")
        
        # 2. Obter playlist
        print("  üî∏ Testando obten√ß√£o de playlist...")
        try:
            obter_playlist_request = streaming_pb2.IdRequest(id="playlist_1")
            playlist = stub.ObterPlaylist(obter_playlist_request)
            print(f"    ‚úÖ Playlist obtida: {playlist.nome}")
        except grpc.RpcError as e:
            print(f"    ‚ùå Erro ao obter playlist: {e.details()}")
        
        # 3. Atualizar playlist
        print("  üî∏ Testando atualiza√ß√£o de playlist...")
        try:
            atualizar_playlist_request = streaming_pb2.AtualizarPlaylistRequest(
                id_playlist="playlist_1",
                nome="Playlist Atualizada",
                musicas=["musica_1", "musica_2", "musica_3"]
            )
            playlist_atualizada = stub.AtualizarPlaylist(atualizar_playlist_request)
            print(f"    ‚úÖ Playlist atualizada: {playlist_atualizada.nome}")
        except grpc.RpcError as e:
            print(f"    ‚ùå Erro ao atualizar playlist: {e.details()}")
        
        # 4. Deletar playlist
        print("  üî∏ Testando dele√ß√£o de playlist...")
        try:
            deletar_playlist_request = streaming_pb2.IdRequest(id="playlist_1")
            resultado = stub.DeletarPlaylist(deletar_playlist_request)
            print(f"    ‚úÖ Playlist deletada: {resultado.success} - {resultado.message}")
        except grpc.RpcError as e:
            print(f"    ‚ùå Erro ao deletar playlist: {e.details()}")
        
        print("\n‚úÖ Todos os testes das opera√ß√µes CRUD foram executados!")
        print("üìù Nota: Este √© um teste de demonstra√ß√£o - os dados n√£o s√£o persistidos.")
        
        return True
        
    except Exception as e:
        print(f"‚ùå Erro geral durante os testes: {e}")
        return False
    
    finally:
        try:
            channel.close()
        except:
            pass

def test_validation_errors():
    """Testa valida√ß√µes de erro"""
    
    print("\nüîç Testando valida√ß√µes de erro...")
    
    try:
        channel = grpc.insecure_channel('localhost:50051')
        stub = streaming_pb2_grpc.StreamingServiceStub(channel)
        
        # Teste 1: Criar usu√°rio com nome vazio
        print("  üî∏ Testando cria√ß√£o de usu√°rio com nome vazio...")
        try:
            criar_request = streaming_pb2.CriarUsuarioRequest(
                nome="",  # Nome vazio
                idade=25
            )
            stub.CriarUsuario(criar_request)
            print("    ‚ùå Deveria ter falhado com nome vazio")
        except grpc.RpcError as e:
            if e.code() == grpc.StatusCode.INVALID_ARGUMENT:
                print(f"    ‚úÖ Valida√ß√£o funcionou: {e.details()}")
            else:
                print(f"    ‚ö†Ô∏è  C√≥digo de erro inesperado: {e.code()}")
        
        # Teste 2: Criar usu√°rio com idade inv√°lida
        print("  üî∏ Testando cria√ß√£o de usu√°rio com idade inv√°lida...")
        try:
            criar_request = streaming_pb2.CriarUsuarioRequest(
                nome="Jo√£o",
                idade=0  # Idade inv√°lida
            )
            stub.CriarUsuario(criar_request)
            print("    ‚ùå Deveria ter falhado com idade inv√°lida")
        except grpc.RpcError as e:
            if e.code() == grpc.StatusCode.INVALID_ARGUMENT:
                print(f"    ‚úÖ Valida√ß√£o funcionou: {e.details()}")
            else:
                print(f"    ‚ö†Ô∏è  C√≥digo de erro inesperado: {e.code()}")
        
        # Teste 3: Buscar usu√°rio inexistente
        print("  üî∏ Testando busca de usu√°rio inexistente...")
        try:
            obter_request = streaming_pb2.IdRequest(id="usuario_inexistente")
            stub.ObterUsuario(obter_request)
            print("    ‚ùå Deveria ter falhado para usu√°rio inexistente")
        except grpc.RpcError as e:
            if e.code() == grpc.StatusCode.NOT_FOUND:
                print(f"    ‚úÖ Valida√ß√£o funcionou: {e.details()}")
            else:
                print(f"    ‚ö†Ô∏è  C√≥digo de erro inesperado: {e.code()}")
        
        print("‚úÖ Testes de valida√ß√£o conclu√≠dos!")
        
    except Exception as e:
        print(f"‚ùå Erro durante testes de valida√ß√£o: {e}")
    
    finally:
        try:
            channel.close()
        except:
            pass
    server_thread.start()
    
    # Aguardar servidor iniciar
    time.sleep(2)
    
    # Conectar ao servidor
    try:
        channel = grpc.insecure_channel('localhost:50051')
        stub = streaming_pb2_grpc.StreamingServiceStub(channel)
        
        print("üß™ Testando opera√ß√µes CRUD do gRPC...")
        
        # Teste 1: Listar usu√°rios
        print("\n1. Testando ListarTodosUsuarios...")
        response = stub.ListarTodosUsuarios(streaming_pb2.Empty())
        print(f"   ‚úÖ Encontrados {len(response.usuarios)} usu√°rios")
        
        # Teste 2: Obter usu√°rio espec√≠fico
        if response.usuarios:
            primeiro_usuario = response.usuarios[0]
            print(f"\n2. Testando ObterUsuario para {primeiro_usuario.id}...")
            user_request = streaming_pb2.UsuarioRequest(id_usuario=primeiro_usuario.id)
            user_response = stub.ObterUsuario(user_request)
            print(f"   ‚úÖ Usu√°rio obtido: {user_response.nome}, {user_response.idade} anos")
        
        # Teste 3: Criar usu√°rio
        print(f"\n3. Testando CriarUsuario...")
        create_request = streaming_pb2.CriarUsuarioRequest(
            nome="Teste Usuario",
            idade=25
        )
        created_user = stub.CriarUsuario(create_request)
        print(f"   ‚úÖ Usu√°rio criado: {created_user.nome} (ID: {created_user.id})")
        
        # Teste 4: Atualizar usu√°rio
        print(f"\n4. Testando AtualizarUsuario...")
        update_request = streaming_pb2.AtualizarUsuarioRequest(
            id_usuario=created_user.id,
            nome="Teste Usuario Atualizado",
            idade=26
        )
        updated_user = stub.AtualizarUsuario(update_request)
        print(f"   ‚úÖ Usu√°rio atualizado: {updated_user.nome}, {updated_user.idade} anos")
        
        # Teste 5: Deletar usu√°rio
        print(f"\n5. Testando DeletarUsuario...")
        delete_request = streaming_pb2.UsuarioRequest(id_usuario=created_user.id)
        delete_response = stub.DeletarUsuario(delete_request)
        print(f"   ‚úÖ Usu√°rio deletado: {delete_response.success} - {delete_response.message}")
        
        # Teste 6: Listar m√∫sicas
        print(f"\n6. Testando ListarTodasMusicas...")
        music_response = stub.ListarTodasMusicas(streaming_pb2.Empty())
        print(f"   ‚úÖ Encontradas {len(music_response.musicas)} m√∫sicas")
        
        # Teste 7: Obter estat√≠sticas
        print(f"\n7. Testando ObterEstatisticas...")
        stats = stub.ObterEstatisticas(streaming_pb2.Empty())
        print(f"   ‚úÖ Estat√≠sticas: {stats.total_usuarios} usu√°rios, {stats.total_musicas} m√∫sicas, {stats.total_playlists} playlists")
        print(f"   üìä M√©dia de m√∫sicas por playlist: {stats.media_musicas_por_playlist:.2f}")
        print(f"   üè∑Ô∏è  Tecnologia: {stats.tecnologia}")
        
        print(f"\nüéâ Todos os testes CRUD do gRPC passaram com sucesso!")
        
    except grpc.RpcError as e:
        print(f"‚ùå Erro gRPC: {e}")
    except Exception as e:
        print(f"‚ùå Erro geral: {e}")
    finally:
        channel.close()

if __name__ == "__main__":
    test_grpc_crud()
